packages/transport/src/chrome-runtime/client-connection-manager.ts

```diff
@@ -0,0 +1,183 @@
+    const connection = {} as Partial<ClientConnection>;
+
+    const name = nameClientConnection(this.label, serviceTypeName);
+    const clientId = name.uuid;
+    const servicePort = chrome.runtime.connect({ includeTlsChannelId: true, name: String(name) });
+
+    /**
+     * Client-specific message handler. Listens for messages on the
+     * page-provided client channel.  Messages are forwarded to the service port
+     * without much validation.
+     *
+     * A reference is kept in case we want to quietly remove the listener later.
+     *
+     * @param ev may be any message transmitted by any untrusted page or other
+     * untrusted script with access to the page-provided message channel
+     */
+    const clientListener = (ev: MessageEvent<TransportMessage | TransportStream>) => {
+      if (!isTransportData(ev.data)) throw Error('Unknown data in transport');
+      const request = ev.data;
+      if (isTransportMessage(request)) {
+        const { requestId, message } = request;
+        servicePort.postMessage({ requestId, message });
+      } // TODO: isTransportStream(request)
+      else throw Error('Unimplemented request kind');
+    };
+
+    /**
+     * Client-specific message handler, listening for return messages from the
+     * extension.  Messages do not necessarily represent a response to a
+     * request.  Messages are forwarded to the page-provided client channel
+     * without much validation.
+     *
+     * The extension may provide a new channel name, to be used as a source for a
+     * stream that gets transferred to the client.
+     *
+     * TODO: response origin?
+     *
+     * @param ev any message transmitted on the client-specific service channel
+     * by any part of the extension runtime, hopefully just the service router
+     */
+    const serviceListener: Parameters<chrome.runtime.PortMessageEvent['addListener']>[0] = (
+      ev: unknown,
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      _: chrome.runtime.Port,
+    ) => {
+      if (!isTransportData(ev)) throw Error('Unknown data in transport');
+      const response = ev;
+      if (isTransportMessage(response)) {
+        const { requestId, message } = response;
+        clientPort.postMessage({ requestId, message } satisfies TransportMessage);
+      } else if (isTransportInitChannel(response)) {
+        const { requestId, channel: name } = response;
+        const sourcePort = chrome.runtime.connect({ includeTlsChannelId: true, name });
+        const stream = new ReadableStream(new ChromeRuntimeStreamSource(sourcePort));
+        clientPort.postMessage({ requestId, stream } satisfies TransportStream, [stream]);
+      } else throw Error('Unimplemented response kind');
+    };
+
+    this.connections.set(
+      clientId,
+      Object.assign(connection, {
+        clientPort,
+        servicePort,
+        clientListener,
+        serviceListener,
+      } satisfies ClientConnection),
+    );
```

Comment:
I find using the spread operator is more common and more readable than object.assign:

```typescript
    this.connections.set(clientId, {
      ...connection,
      clientPort,
      servicePort,
      clientListener,
      serviceListener,
    } satisfies ClientConnection);
```